pep-0636.rst

```diff
@@ -0,0 +1,382 @@
+the same time we get better input validation, and we will not be getting into that
+branch if the command entered by the user is ``"go figure!"`` instead of an direction.
+
+We could try to get the best of both worlds doing the following (I'll omit the aliased
+version without "go" for brevity)::
+
+   match command.split():
+       case ["go", ("north" | "south" | "east" | "west")]:
+           current_room = current_room.neighbor(...)
+           # how do I know which direction to go?
+
+This code is a single branch, and it verifies that the word after "go" is really a
+direction. But the code moving the player around needs to know which one was chosen and
+has no way to do so. What we need is a pattern that behaves like the or pattern but at
+the same time does a capture. We can do so with a **walrus pattern**::
+
+   match command.split():
+       case ["go", direction := ("north" | "south" | "east" | "west")]:
+           current_room = current_room.neighbor(direction)
+
+The walrus pattern (named like that because the ``:=`` operator looks like a sideways
+walrus) matches whatever pattern is on its right hand side, but also binds the value to
+a name.
+
+Conditional pattern matching
+----------------------------
+
+The patterns we have explored above can do some powerful data filtering, but sometimes
+you may wish for the full power of a boolean expression. Let's say that you would actually
+like to allow a "go" command only in a restricted set of directions based on the possible
+exits from the current_room. We can achieve that by adding a **guard** to our
+case-clause. Guards consist of the ``if`` keyword followed by any expression::
+
+   match command.split():
+       case ["go", direction] if direction in current_room.exits:
+           current_room = current_room.neighbor(direction)
+       case ["go", _]:
+           print("Sorry, you can't go that way") 
+
+The guard is not part of the pattern, it's part of the case clause. It's only checked if
+the pattern matches, and after all the pattern variables have been bound (that's why the
+condition can use the ``direction`` variable in the example above). If the pattern
+matches and the condition is truthy, the body of the case clause runs normally. If the
+pattern matches but the condition is falsy, the match statement proceeds to check the
+next ``case`` clause as if the pattern hadn't matched (with the possible side-effect of
+having already bound some variables).
+
+The sequence of these steps must be considered carefully when combining or-patterns and
+guards. If you have ``case [x, 100] | [0, x] if x > 10`` and your subject is 
+``[0, 100]``, the clause will be skipped. This happens because:
+
+* The or-pattern finds the first alternative that matches the subject, which happens to
+  be ``[x, 100]``
+* ``x`` is bound to 0
+* The condition x > 10 is checked. Given that it's false, the whole case clause is
+  skipped. The ``[0, x]`` pattern is never attempted.
```

Comment:
That's a clever example, but (again) it feels too advanced for a tutorial of this level. I doubt many people will ever run into this -- I note that you couldn't think of an example in the context of your adventure game.

Reply:
I may end up removing it.